### DISCUSSION TITLE
GType links are broken in documentation

### DISCUSSION THREAD

**Discussion by AttilaTheFun** [libvips/libvips#2413] [Q&A]
I was trying to read more about GTypes but the links in the documentation were broken. Here is one such example:
https://libvips.github.io/libvips/usr/share/gtk-doc/html/gobjectgobject-Type-Information.html#GType

After a bit of googling, I believe this is what it was supposed to be linking to:
http://www.gnu.org/software/g-golf/manual/html_node/Type-Information.html

Is that the case? And if so, can we repair those links? Thanks!

**Comment by jcupitt** [comment]
Hi @AttilaTheFun, 

This is a gtk-doc feature. In gtk-doc you mark up names from other packages with a % sign, for example:

```C

/**
 * vips_area_new: (constructor)
 * @free_fn: (scope async): @data will be freed with this function
 * @data: data will be freed with this function
 *
 * A VipsArea wraps a chunk of memory. It adds reference counting and a free
 * function. It also keeps a count and a %GType, so the area can be an array.
 * ...
```

https://github.com/libvips/libvips/blob/master/libvips/iofuncs/type.c#L245

The doc generator will link these names to the docs for the other packages on this local machine. For the on-line docs, there's only libvips, so these inter-package links don't work. They will link correctly if you build the docs yourself and have a local copy of the glib docs.

GType is part of GObject, which is the object system for glib. The official docs are here:

https://docs.gtk.org/gobject/

And GType is here:

https://docs.gtk.org/gobject/alias.Type.html

There's a background chapter in the GObject docs introducing types here:

https://docs.gtk.org/gobject/concepts.html

**Reply by AttilaTheFun** [reply]
Ah that makes sense! That said, most library consumers are probably going to be viewing the docs online rather than building them locally. It looks like there are only five %GType references in the repo right now - maybe it's worth replacing them with https links which would work for everyone?
https://github.com/libvips/libvips/search?q=%25GType

**Reply by jcupitt** [reply]
True, though gtk-doc will link other things too -- for example, function references such as `g_type_name()` will be linked to their docs. And there are things other than gtype:

```
$ grep "^ \* .*%" */*.c | wc
   1044   10206   79744
```

ie. look for `%` characters in comments. There's `GValue`, `%gchararray`, etc.

We could build and host a copy of the glib docs, that way we'd get cross-linking done for us, but that seems like duplication of effort since the best place for glib docs is obviously the glib site.

**Reply by jcupitt** [reply]
Another alternative would be patch the generated HTML when the API docs are built.

The `/API` part of the libvips site is built by this script:

https://github.com/libvips/libvips/blob/gh-pages/gen-api.rb

Perhaps we could spot things like `GType` and `GValue` there? Maybe even the `g_*()` function calls.
